File: go/bones.go
// bones.go - a simple library for assembling valid HTML content
//
// Inspiration "Behind the Code: Avoiding Spaghetti" by Jason Grosman
// http://www.npr.org/blogs/inside/2011/02/02/126312263/behind-the-code-avoiding-spaghetti-html
//
// HTML markup reference was http://en.wikipedia.org/wiki/HTML_element#HTML401
//
// author R. S. Doiel, <rsdoiel@gmail.com>
//
// copyright (c) 2011 all rights reserved
//
// Released under New the BSD License.
// See: http://opensource.org/licenses/bsd-license.php
//
package bones

import (
	"strings"
	"bytes"
	"fmt"
)

const (
	VERSION = `0.0.1-a`
)

// Assemble a map of attribute names and values
func AssembleAttributes(attributes map[string]string) string {
	out := new(bytes.Buffer)
	for key, value := range attributes {
		out.Write([]byte(fmt.Sprintf(` %s="%s"`, key, value)))
	}
	return out.String()
}

// Assemble a generic tag from name, the innerHTML and any attributes
func AssembleTag(name string, innerHTML string, attributes map[string]string) string {
	switch {
	case name == "html" && attributes == nil && innerHTML == "":
		return strings.Join([]string{"<", name, "></", name, ">",}, "")
	case name == "html" && attributes != nil && innerHTML == "":
		return strings.Join([]string{"<", name, AssembleAttributes(attributes), "></",name,">",}, "")
	case name == "p" && attributes == nil && innerHTML == "":
		return strings.Join([]string{"<", name, ">"}, "")
	case name == "p" && attributes != nil && innerHTML == "":
		return strings.Join([]string{"<", name, AssembleAttributes(attributes), ">"}, "")
	case attributes == nil && innerHTML != "":
		return strings.Join([]string{"<", name, ">", innerHTML, "</", name, ">"}, "")
	case attributes == nil && innerHTML == "":
		return strings.Join([]string{"<", name, "/>"}, "")
	case innerHTML != "":
		return strings.Join([]string{"<", name, AssembleAttributes(attributes), ">", innerHTML, "</", name, ">"}, "")
	case innerHTML == "":
		return strings.Join([]string{"<", name, AssembleAttributes(attributes), " />"}, "")
	}
	return ""
}

// Outer HTML element including DOCTYPE prefix
func Html (innerHTML string, attributes map[string]string) (string) {
	return fmt.Sprintf("%s\n%s","<!DOCTYPE html>", AssembleTag("html", innerHTML, nil))
}

// Head element
func Head (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("head", innerHTML, attributes)
}

// Title element
func Title (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("title", innerHTML, attributes)
}

// Base element
func Base (href string, attributes map[string]string) (string) {
        attributes["href"] = href
	return AssembleTag("base", "", attributes)
}

func Link (attributes map[string]string) (string) {
	return AssembleTag("link", "", attributes)
}

func LinkCSS (href string, attributes map[string]string) (string) {
	attributes["href"] = href
	return AssembleTag("link", "", attributes)
}

func Meta (attributes map[string]string) (string) {
	return AssembleTag("meta","",attributes)
}

func Object (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("object", innerHTML, attributes)
}

func Style (CSS string, attributes map[string]string) (string) {
	return AssembleTag("style",CSS,attributes)
}

func Body (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("body", innerHTML, attributes)
}

func H1 (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("h1",innerHTML, attributes)
}

func H2 (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("h2", innerHTML, attributes)
}

func H3 (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("h3", innerHTML, attributes)
}

func H4 (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("h4", innerHTML, attributes)
}

func H5 (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("h5", innerHTML, attributes)
}

func H6 (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("h6", innerHTML, attributes)
}

func P (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("p",innerHTML, attributes)
}

func A (innerHTML string, href string, attributes map[string]string) (string) {
	attributes["href"] = href
	return AssembleTag("a", innerHTML, attributes)
}

func Img (url string, attributes map[string]string) (string) {
	attributes["src"] = url
	return AssembleTag("img", "", attributes)
}

func Ul (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("ul", innerHTML, attributes)
}

func Ol (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("ol", innerHTML, attributes)
}

func Li (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("li", innerHTML, attributes)
}

func Dl (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("dl", innerHTML, attributes)
}
 
func Dt (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("dt", innerHTML, attributes)
}

func Dd (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("dd", innerHTML, attributes)
}

func Table (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("table", innerHTML, attributes)
}

func Th (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("th", innerHTML, attributes)
}

func Tr (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("tr", innerHTML, attributes)
}

func Col (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("col", innerHTML, attributes)
}

func Colgroup (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("colgroup", innerHTML, attributes)
}

func THead (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("thead", innerHTML, attributes)
}

func TBody (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("tbody", innerHTML, attributes)
}

func TFoot (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("tfoot", innerHTML, attributes)
}

func Td (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("td", innerHTML, attributes)
}

func Form (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("form", innerHTML, attributes)
}

func Label(innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("label", innerHTML, attributes)
}

func Input (name string, value string, attributes map[string]string) (string) {
	attributes["name"] = name
	attributes["value"] = value
	return AssembleTag("input", "", attributes)
}

func Textarea (name string, value string, attributes map[string]string) (string) {
	attributes["name"] = name
	return AssembleTag("textarea", value, attributes)
}

func Select (name string, innerHTML string, attributes map[string]string) (string) {
	attributes["name"] = name
	return AssembleTag("select", innerHTML, attributes)
}

func Option (value string, label string, attributes map[string]string) (string) {
	attributes["value"] = value
	return AssembleTag("option", label, attributes)
}

func Optgroup (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("optgroup", innerHTML, attributes)
}

func Button (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("button", innerHTML, attributes)
}

func Fieldset (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("fieldset", innerHTML, attributes)
}

func Legend (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("legend", innerHTML, attributes)
}

func Script (url string, code string, attributes map[string]string) (string) {
	if url != "" {
		attributes["src"] = url
		return AssembleTag("script", "", attributes)
	}
	return AssembleTag("script", code, attributes)
}

func NoScript (innerHTML string) (string) {
	return AssembleTag("noscript", innerHTML, nil)
}

func Address(innerHTML string) (string) {
	return AssembleTag("address", innerHTML, nil)
}

func Blockquote(innerHTML string) (string) {
	return AssembleTag("backquote", innerHTML, nil)
}

func Del(innerHTML string) (string) {
	return AssembleTag("del", innerHTML, nil)
}

func Hr(attributes map[string]string) (string) {
	return AssembleTag("hr", "", attributes)
}

func Ins(innerHTML string) (string) {
	return AssembleTag("ins", innerHTML, nil)
}

func Pre(innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("pre", innerHTML, attributes)
}

func Div (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("div", innerHTML, attributes)
}

func Span (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("span", innerHTML, attributes)
}

func Abbr (abbreviation string, full_form string, attributes map[string]string) (string) {
	attributes["title"] = full_form
	return AssembleTag("abbr", abbreviation, attributes)
}

func Acronym (acronym string, full_form string, attributes map[string]string) (string) {
	attributes["title"] = full_form
	return AssembleTag("acronym", acronym, attributes)
}

func Dfn (innerHTML string) (string) {
	return AssembleTag("dfn", innerHTML, nil)
}

func Em (innerHTML string) (string) {
	return AssembleTag("em", innerHTML, nil)
}

func Strong (innerHTML string) (string) {
	return AssembleTag("strong", innerHTML, nil)
}

func Code (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("code", innerHTML, attributes)
}

func Samp (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("samp", innerHTML, attributes)
}

func Kbr (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("kbr", innerHTML, attributes)
}

func Var (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("var", innerHTML, attributes)
}

func B (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("b", innerHTML, attributes)
}

func I (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("i", innerHTML, attributes)
}

func Big (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("big", innerHTML, attributes)
}

func Small (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("small", innerHTML, attributes)
}

func Sub (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("sub", innerHTML, attributes)
}

func Sup (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("sup", innerHTML, attributes)
}

func Tt (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("tt", innerHTML, attributes)
}

func Br (attributes map[string]string) (string) {
	return AssembleTag("br", "", attributes)
}

func Bdo (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("bdo", innerHTML, attributes)
}

func Cite (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("cite", innerHTML, attributes)
}

func Q (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("q", innerHTML, attributes)
}

func Area (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("area", innerHTML, attributes)
}

func Map (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("map", innerHTML, attributes)
}

func Frame (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("frame", innerHTML, attributes)
}

func NoFrame (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("noframe", innerHTML, attributes)
}

func IFrame (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("iframe", innerHTML, attributes)
}

func HGroup (innerHTML string, attributes map[string]string) (string) {
	return AssembleTag("hgroup", innerHTML, attributes)
}

File: js/bones.js
/**
 * bones is a simple library for assembling valid HTML. It is available in
 * three languages - Go, JavaScript and PHP.
 *
 * Inspiration "Behind the Code: Avoiding Spaghetti" by Jason Grosman
 * http://www.npr.org/blogs/inside/2011/02/02/126312263/behind-the-code-avoiding-spaghetti-html
 *
 * HTML markup reference was http://en.wikipedia.org/wiki/HTML_element#HTML401
 *
 * author R. S. Doiel, <rsdoiel@gmail.com>
 *
 * copyright (c) 2011 all rights reserved
 *
 * Released under New the BSD License.
 * See: http://opensource.org/licenses/bsd-license.php
 *
 * Notes: this should work in the browser or under NodeJS, test-bones.js is designed to run under NodeJS
 */

var bones = {
	/**
	 * Factory - attach the bones functions to an object.
	 * @params self - the object to have the functions attached to.
	 */
	Factory : function (self) {
		var ky;

		if (self === undefined) {
			var self = {};
		}
		for (ky in this) {
			if (typeof this[ky] === 'function' &&
				self[ky] === undefined) {
				self[ky] = this[ky];
			}
		}
		return self;
	}, // End: Factory()

	/**
	 * trim - a convenience function to trim the whitespace from the 
	 * start and end of a string.
	 * @param s - the string to trim
	 * @return a trimmed string
	 */
	trim : function (s) {
		return s.replace(/\s+$/,'').replace(/^\s+/,'');
	},
	
	/**
	 * assemble_attributes - formats a valid set of attribute strings.
	 * @param attributes - either an associative array or formatted string.
	 * @return a string representation of the attributes.
	 */
	assemble_attributes : function (attributes) {
		var attr = '', key;

		if (attributes === undefined || attributes == null) {
			return '';
		} else if (typeof attributes === 'object') {
			for (key in attributes) {
                if (typeof key !== "function") {
					value = attributes[key];
				    if (attributes[key] !== null) {
					    attr += ' ' + key + '="' + this.trim(attributes[key]) + '"';
				    } else {
					    attr += ' ' + key;
				    }
                }
			}
			return ' ' + this.trim(attr);
		} else {
			return ' ' + this.trim(attributes);
		}
	}, /* END: assemble_attributes() */

	/**
	 * disassemble_attributes - parse an attribute string (e.g. 'class="myclass" id="fred"') into an associative array.
	 * @param attributes_string - a list of attributes in string form.
	 * @return an object representation of the attributes.
	 */
	disassemble_attributes : function (attributes_string) {
		var in_quote = false,
				key = false,
				key_start = 0,
				value_start = 0,
				attributes = {}, pos = 0, chr = ' ';

		str = this.trim(attributes_string);
		for(pos = 0; pos < str.Length; pos += 1) {
			chr = str.substr(pos, 1);
			if (in_quote) {
				// Have we exited quote
				if (chr == in_quote) {
					attributes[key] = str.substr(value_start, pos - value_start);
					in_quote = false;
					value = 0;
					key = false;
					key_start = pos + 1;
				} else if (chr == '\\' && str.substr(pos + 1, 1) == in_quote) {
					pos += 1;
				}
			} else if (chr == '"' || chr == "'") {
				in_quote = chr;
				value_start = pos + 1;
			} else if (chr == '=') {
				key = this.trim(str.substr(key_start, pos - key_start));
			}
		}
		return attributes;
	}, /* END: disassemble_attributes() */
	
	/**
	 * Html - for the outer HTML and Doctype wrapper for an HTML page.
	 * @param innerHTML - usually a head and body tag with their contents.
	 * @param attributes - usually something like lange="en"
	 * @return a string representation of the HTML page.
	 */
	Html : function (innerHTML, attributes) {
		if (attributes === undefined) {
			attributes = 'lang="en"';
		}
		return '<!DOCTYPE html>' + "\n" +
		'<html' + this.assemble_attributes(attributes) + '>' + "\n" + innerHTML + '</html>';
	}, /* END: Html() */
	
	/**
	 * Head - render a html head element.
	 * @param innerHTML - the contents of head tag
	 * @param attributes - e.g. a key/value pair of attribute name and value or string like class="myclass"
	 * @return a string representation of the head block
	 */
	Head : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<head' + this.assemble_attributes(attributes) + '>' + "\n" + innerHTML + '</head>' + "\n";
	},
	
	/**
	 * Title - render an html title element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Title : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<title' + this.assemble_attributes(attributes) + '>' + innerHTML + '</title>';
	},
	
	/**
	 * Link - an html link element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Link : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<link' + this.assemble_attributes(attributes) + ' />';
	},
	
	/**
	 * Body - render an html body element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Body : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<body' + this.assemble_attributes(attributes) + '>' + "\n" + innerHTML + '</body>';
	},
	
	/**
	 * H1 - render an html h1 element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	H1 : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<h1' + this.assemble_attributes(attributes) + '>' + innerHTML + '</h1>';
	},
	
	/**
	 * H2 - render an html h2 element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	H2 : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<h2' + this.assemble_attributes(attributes) + '>' + innerHTML + '</h2>';
	},
	
	/**
	 * H3 - render an html h3 element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	H3 : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<h3' + this.assemble_attributes(attributes) + '>' + innerHTML + '</h3>';
	},
	
	/**
	 * H4 - render an html h4 element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	H4 : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<h4' + this.assemble_attributes(attributes) + '>' + innerHTML + '</h4>';
	},
	
	/**
	 * H5 - render an html h5 element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	H5 : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<h5' + this.assemble_attributes(attributes) + '>' + innerHTML + '</h5>';
	},
	
	/**
	 * H6 - render an html h6 element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	H6 : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<h6' + this.assemble_attributes(attributes) + '>' + innerHTML + '</h6>';
	},
	
	/**
	 * P - render an html paragraph element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	P : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			return '<p' + this.assemble_attributes(attributes) + '>';
		} else {
			return '<p' + this.assemble_attributes(attributes) + '>' + innerHTML + '</p>';
		}
	},
	
	/**
	 * A - render an html anchor element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	A : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		if (typeof attributes === "string" && attributes.indexOf("=") < 0) {
			attributes = {'href' : attributes};			
		}
		return '<a' + this.assemble_attributes(attributes) + '>' + innerHTML + '</a>';
	},
	
	/**
	 * Ul - render an html un-order list element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Ul : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<ul' + this.assemble_attributes(attributes) + '>' + innerHTML + '</ul>';
	},
	
	
	/**
	 * Ol - render an html order list element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Ol : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<ol' + this.assemble_attributes(attributes) + '>' + innerHTML + '</ol>';
	},
	
	/**
	 * Li - render an html list item element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Li : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<li' + this.assemble_attributes(attributes) + '>' + innerHTML + '</li>';
	},
	
	/**
	 * Dl - render an html definition list element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Dl : function (innerHTML, attributes) { 
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<dl' + this.assemble_attributes(attributes) + '>' + innerHTML + '</dl>';
	},
	
	/**
	 * Dt - render an html definition term element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Dt : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<dt' + this.assemble_attributes(attributes) + '>' + innerHTML + '</dt>';
	},
	
	/**
	 * Dd - render an html definition defined element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Dd : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<dd' + this.assemble_attributes(attributes) + '>' + innerHTML + '</dd>';
	},
	
	/**
	 * Table - render an html a table element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Table : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<table' + this.assemble_attributes(attributes) + '>' + innerHTML + '</table>';
	},
	
	/**
	 * Th - render an html table header element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Th : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<th' + this.assemble_attributes(attributes) + '>' + 
			innerHTML + '</th>';
	},
	
	/**
	 * Tr - render an html table row element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Tr : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<tr' + this.assemble_attributes(attributes) + '>' + 
			innerHTML + '</tr>';
	},
	
	/**
	 * Td - render an html table data element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Td : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<td' + this.assemble_attributes(attributes) + '>' + innerHTML + '</td>';
	},
	
	/**
	 * Form - render an html form element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Form : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<form' + this.assemble_attributes(attributes) + '>' + innerHTML + '</form>';
	},
	
	/**
	 * Input - render an html input element
	 * @param name - the name of the input field
	 * @param value - the default value of the input field
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Input : function (name, value, attributes) {
		if (name === undefined) {
			name = '';
		}
		if (value === undefined) {
			value = '';
		}
		return '<input name="' + name + '" value="' + value + '"' + this.assemble_attributes(attributes) + ' />';
	},
	
	/**
	 * Button - create an input button 
	 * @param name - the name of the button
	 * @param value - the value of the button
	 * @attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Button : function (name, value, attributes) {
		var attr = this.disassemble_attributes(this.assemble_attributes(attributes));
		attr.type = 'button';
		return this.Input(name, value, attr);
	},
	
	/**
	 * Textarea - render an html textarea element
	 * @param name - the name of a textarea
	 * @param value - the default value of the textarea
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Textarea : function (name, value, attributes) {
		if (value === undefined) {
			value = '';
		}
		return '<textarea name="' + name + '"' + this.assemble_attributes(attributes) + '>' + value + '</textarea>';
	},
	
	/**
	 * Select - an html select element
	 * @param name - name of the select element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Select : function (name, innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		if (name === undefined) {
			name = '';
		}
		return '<select name="' + name + '"' + this.assemble_attributes(attributes) + '>' + innerHTML + '</select>';
	},
	
	/**
	 * Option - an html option element
	 * @param value - a value for the element
	 * @param label - for the option
	 * @return a string representation of the element
	 */
	Option : function (value, label, attributes) {
		if (value === undefined) {
			value = '';
		}
		if (label === undefined) {
			label = '';
		}
		return '<option value="' + value + '"' + this.assemble_attributes(attributes) + '>' + label + '</option>';
	},
	
	/**
	 * Label - an html label element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or hash of key/values representing attributes
	 * @return a string representation of the element
	 */
	Label : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<label' + this.assemble_attributes(attributes) + '>' + innerHTML + '</label>';
	},
	
	/**
	 * Script - an html script element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or hash of key/values representing attributes
	 * @return a string representation of the element
	 */
	Script : function (url, code, attributes) {
		if (url !== undefined) {
			return '<script type="JavaScript" rel="text/javascript" src="' + url + '"' + this.assemble_attributes(attributes) + '></script>';
		} else {
			return '<script type="JavaScript" rel="text/javascript"' + this.assemble_attributes(attributes) + '>' + code + '</script>';
		}
	},
	
	/**
	 * Pre - an html pre element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or hash of key/values representing attributes
	 * @return a string representation of the element
	 */
	Pre : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<pre' + this.assemble_attributes(attributes) + '>' + innerHTML + '</pre>';
	},
	
	/**
	 * DemoCode - an html composite of code and pre tags
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or hash of key/values representing attributes
	 * @return a string representation of the element
	 */
	DemoCode : function (code_source, attributes) {
		if (code_source === undefined) {
			code_source = '';
		}
		return '<code' + this.assemble_attributes(attributes) + '>' + Pre( "\n" + code_source + "\n") + '</code>';
	},
	
	/**
	 * Div - an html div element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or hash of key/values representing attributes
	 * @return a string representation of the element
	 */
	Div : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<div' + this.assemble_attributes(attributes) + '>' + innerHTML + '</div>';
	},
	
	/**
	 * Span - an html span element
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or hash of key/values representing attributes
	 * @return a string representation of the element
	 */
	Span : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<span' + this.assemble_attributes(attributes) + '>' + innerHTML + '</span>';
	},

	/**
	 * Menu - an HTML5 Menu element
	 * @param label
	 * @param innerHTML
	 * @param attributes
	 * @return a string representation of the element
	 */
	Menu : function (label, innerHTML, attributes) {
		return '<menu label="' + label + '"' + this.assemble_attributes(attributes) + '>' + innerHTML + '</menu>';
	},
	
	/**
	 * Img - image element
	 * @param src
	 * @param attributes
	 * @return a string representation of the element
	 */
	Img : function (src, attributes) {
		return '<img src="' + src + '"' + this.assemble_attributes(attributes) + '/>';
	},
	
	/**
	 * Center - render a center tag
	 * @param innerHTML - the contents of tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Center : function (innerHTML, attributes) {
		if (innerHTML === undefined) {
			innerHTML = '';
		}
		return '<center' + this.assemble_attributes(attributes) + '>' + innerHTML + '</center>';
	},
	/**
	 * Br - render a br tag
	 * @param attributes - a string or object of key/values representing attributes
	 * @return a string representation of the element
	 */
	Br : function (innerHTML, attributes) {
		return '<br' + this.assemble_attributes(attributes) + '/>';
	}
};

// Defined some exports if running under NodeJS
try {
	if (exports !== undefined) {
		exports.Factory = bones.Factory;
		exports.trim = bones.trim;
		exports.assemble_attributes = bones.assemble_attributes;
		exports.Html = bones.Html;
		exports.Head = bones.Head;
		exports.Title = bones.Title;
		exports.Link = bones.Link;
		exports.Body = bones.Body;
		exports.H1 = bones.H1;
		exports.H2 = bones.H2;
		exports.H3 = bones.H3;
		exports.H4 = bones.H4;
		exports.H5 = bones.H5;
		exports.H6 = bones.H6;
		exports.P = bones.P;
		exports.Br = bones.Br;
		exports.A = bones.A;
		exports.Ul = bones.Ul;
		exports.Ol = bones.Ol;
		exports.Li = bones.Li;
		exports.Dl = bones.Dl;
		exports.Dt = bones.Dt;
		exports.Dd = bones.Dd;
		exports.Table = bones.Table;
		exports.Th = bones.Th;
		exports.Tr = bones.Tr;
		exports.Td = bones.Td;
		exports.Form = bones.Form;
		exports.Input = bones.Input;
		exports.Textarea = bones.Textarea;
		exports.Select = bones.Select;
		exports.Option = bones.Option;
		exports.Label = bones.Label;
		exports.Script = bones.Script;
		exports.Pre = bones.Pre;
		exports.DemoCode = bones.DemoCode;
		exports.Div = bones.Div;
		exports.Span = bones.Span;
		exports.Menu = bones.Menu;
		exports.Img = bones.Img;
		exports.Center = bones.Center;
	} // END: defining exports
} catch(err) {
	// Ignore if exports not supported. E.g. in the browser
};

